agent/eventhandler/task_handler.go

```diff
@@ -15,63 +15,80 @@ package eventhandler
 
 import (
 	"container/list"
+	"sync"
 	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the TaskHandler
+const concurrentEventCalls = 3
 
-func init() {
-	handler = newTaskHandler()
+type eventList struct {
+	*list.List            // list of *sendableEvents
+	sending    bool       // whether the list is already being handled
+	mu         sync.Mutex // Locks both the list and sending bool
+}
+
+type TaskHandler struct {
+	submitSemaphore utils.Semaphore       // Semaphore on the number of tasks that may be handled at once
+	taskMap         map[string]*eventList // arn:*eventList map so events may be serialized per task
+	taskMapMu       sync.RWMutex          // Lock for the taskMap
+}
+
+func NewTaskHandler() *TaskHandler {
+	taskMap := make(map[string]*eventList)
+	submitSemaphore := utils.NewSemaphore(concurrentEventCalls)
+
+	return &TaskHandler{
+		taskMap:         taskMap,
+		submitSemaphore: submitSemaphore,
+	}
 }
 
 // AddTaskEvent queues up a state change for sending using the given client.
-func AddTaskEvent(change api.TaskStateChange, client api.ECSClient) {
-	addEvent(newSendableTaskEvent(change), client)
+func (handler *TaskHandler) AddTaskEvent(change api.TaskStateChange, client api.ECSClient) {
+	handler.addEvent(newSendableTaskEvent(change), client)
 }
 
 // AddContainerEvent queues up a state change for sending using the given client.
-func AddContainerEvent(change api.ContainerStateChange, client api.ECSClient) {
-	addEvent(newSendableContainerEvent(change), client)
+func (handler *TaskHandler) AddContainerEvent(change api.ContainerStateChange, client api.ECSClient) {
+	handler.addEvent(newSendableContainerEvent(change), client)
 }
 
 // Prepares a given event to be sent by adding it to the handler's appropriate
 // eventList
-func addEvent(change *sendableEvent, client api.ECSClient) {
-	var taskList *eventList
-	var preexisting bool
+func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
+
 	log.Info("Adding event", "change", change)
 
-	// TaskEvents lock scope
-	func() {
-		handler.Lock()
-		defer handler.Unlock()
+	handler.taskMapMu.Lock()
+	defer handler.taskMapMu.Unlock()
```

Comment:
This `defer`'ed `handler.taskMapMu.Unlock()` should be moved at the end of the operation on the map to prevent contention while waiting on the rest of the method instead of `defer`-ing to the end of the function scope.